iFixit/toolbox
# Design validity

This pull adds invalid designs to our toolbox components `Textarea`, `TextField`, `Checkbox`, and `Radio`.

<img src="https://user-images.githubusercontent.com/4934193/30122511-88d5183e-92e3-11e7-8ff9-c57e43452012.gif" width="514">

# Textarea, TextField

## States

❌ Means the property's value doesn't matter for that state.

| Show Validity | Valid | Show Validation Message | UI |
| - | - | - | - |
| False | ❌ | ❌ | ![image](https://user-images.githubusercontent.com/4934193/30122785-98b12ec2-92e4-11e7-9b03-b1696d984b93.png)
| True | False | False | ![image](https://user-images.githubusercontent.com/4934193/30122850-c1461d48-92e4-11e7-8075-7a8f5df47bb7.png)
| True | False | False | ![image](https://user-images.githubusercontent.com/4934193/30122939-215fbe82-92e5-11e7-85a5-38a4aacfa143.png)
| True | True | ❌ | ![image](https://user-images.githubusercontent.com/4934193/30122902-f7d2a548-92e4-11e7-9927-25993356c4a9.png)

## UX

### Submitting

- If a field is invalid
  - Submit event is cancelled
  - `Valid` is set to `false`
  - `Show Validity` is set to `true`
  - Validation Message is updated

### Clicking Validity Icon

- If `Valid` is `false`
  - `Show Validation Message` is toggled
  - Validation Message is updated

### Correctly filling out an invalidated field

- `Valid` is set to `true`
- `Show Validation Message` is set to `false`
- When the field is unfocused, `Show Validity` is set to `false`

# Checkbox, Radio

_Note: A checkbox or radio can only be "invalid" if it's required to be checked, and is unchecked. This makes the UI and UX much simpler since there can only be **1** type of invalidity, and **1** way to validate it (by checking)._

## States

❌ Means the property's value doesn't matter for that state.

| Show Validity | Checked | UI |
| - | - | - |
| False | False | ![image](https://user-images.githubusercontent.com/4934193/30124550-5c1ad6b0-92ea-11e7-9de1-b1bc05cd73ef.png)
| True | False | ![image](https://user-images.githubusercontent.com/4934193/30124585-7b78701c-92ea-11e7-9411-3f8cbb5d2d5f.png)
| ❌ | True | ![image](https://user-images.githubusercontent.com/4934193/30124639-9b37563e-92ea-11e7-925b-0dc74c30b0d2.png)

## UX

### Submitting

- If a field is invalid
  - Submit event is cancelled
  - `Show Validity` is set to `true`

### Checking an invalidated input

  - `Show Validity` is set to `false`
  - `Checked` is set to `true`

# QA

---

- Repeat this process for
  - https://deploy-preview-13--toolbox.netlify.com/#!/Textarea/9
  - https://deploy-preview-13--toolbox.netlify.com/#!/TextField/9
- Assert there's an invalid icon
- Assert that clicking on the invalid icon shows the validation message
- Assert clicking on the invalid icon again removes the validation message
- Wash, rinse, repeat

---

- Repeat this process for
  - https://deploy-preview-13--toolbox.netlify.com/#!/Textarea/11
  - https://deploy-preview-13--toolbox.netlify.com/#!/TextField/11
- Assert there's a valid icon
- Assert that clicking on the valid icon doesn't do anything

---

- Repeat this process for
  - https://deploy-preview-13--toolbox.netlify.com/#!/Checkbox/9
  - https://deploy-preview-13--toolbox.netlify.com/#!/Radio/9
- Assert there's an invalid design for the inputs

---

- Run this in your terminal (not on cominor)
  - git clone https://github.com/iFixit/formy.git
  - cd formy/
  - git checkout design-validity
  - npm install
  - npm start
- A browser window should open with the formy test page
- Scroll to the "Form with validity checking" section
- Assert that the form follows the UX sections written in this pull

--- 

- Assert consistency across browsers and devices

Closes https://github.com/iFixit/design/issues/109
